src/Metadata/Extractor/AbstractExtractor.php

```diff
@@ -52,4 +57,64 @@ public function getResources(): array
      * Extracts metadata from a given path.
      */
     abstract protected function extractPath(string $path);
+
+    /**
+     * Recursively replaces placeholders with the service container parameters.
+     *
+     * @see https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Routing/Router.php
+     * @copyright (c) Fabien Potencier <fabien@symfony.com>
+     *
+     * @param mixed $value The source which might contain "%placeholders%"
+     *
+     * @return mixed The source with the placeholders replaced by the container
+     *               parameters. Arrays are resolved recursively.
+     *
+     * @throws ParameterNotFoundException When a placeholder does not exist as a container parameter
+     * @throws RuntimeException           When a container value is not a string or a numeric value
+     */
+    protected function resolve($value)
+    {
+        if (null === $this->container) {
+            return $value;
+        }
+
+        if (\is_array($value)) {
+            foreach ($value as $key => $val) {
+                $value[$key] = $this->resolve($val);
+            }
+
+            return $value;
+        }
+
+        if (!\is_string($value)) {
+            return $value;
+        }
+
+        $escapedValue = preg_replace_callback('/%%|%([^%\s]++)%/', function ($match) use ($value) {
+            // skip %%
+            if (!isset($match[1])) {
+                return '%%';
+            }
+
+            if (preg_match('/^env\(\w+\)$/', $match[1])) {
+                throw new RuntimeException(sprintf('Using "%%%s%%" is not allowed in routing configuration.', $match[1]));
+            }
+
+            if ($this->container instanceof SymfonyContainerInterface) {
+                $resolved = $this->container->getParameter($match[1]);
+            } else {
+                $resolved = $this->container->get($match[1]);
```

Comment:
this will fail if no container is provided, maybe an empty container could be set in the constructor in this scenario